### PR TITLE
Proof-of-stakiness based on block header

### LINKED ISSUE #2349: Check proof-of-stakiness based on block header
In several places, we do a check similar to `com.consensus == ConsensusType.POS` to determine block validation rules - similar to #2322 the block should be checked based on its own header information rather than a global com state.

### PR DESCRIPTION
**Pull request by jangko**
fix #2349

### CODE REVIEW

File: nimbus/evm/state.nim [hunk 1]
```diff
@@ -241,8 +242,15 @@ func blockNumber*(vmState: BaseVMState): BlockNumber =
   # and not head.number
   vmState.parent.number + 1
 
-func difficultyOrPrevRandao*(vmState: BaseVMState): UInt256 =
-  if vmState.com.consensus == ConsensusType.POS:
+proc proofOfStake(vmState: BaseVMState): bool =
+  vmState.com.proofOfStake(Header(
+    number: vmState.blockNumber,
+    parentHash: vmState.blockCtx.parentHash,
+    difficulty: vmState.blockCtx.difficulty,
```
Comment (tersec) [inline nimbus/evm/state.nim:249]: isn't this always 0 in PoS?

File: nimbus/common/hardforks.nim [hunk 2]
```diff
@@ -177,8 +168,6 @@ type
 
     terminalTotalDifficulty*: Opt[UInt256]
     terminalTotalDifficultyPassed*: Opt[bool]
```
Comment (tersec) [inline nimbus/common/hardforks.nim:170]: Not entirely clear what the value of the TTD is at this point.

It was a transition mechanism but `nimbus-eth1` will never have to do a merge, only trust that the blocks provided are indeed the canonical chain.

So is it for passing test vectors, maybe? If they check if proper TTD checks are implemented.

File: nimbus/common/common.nim [hunk 3]
```diff
@@ -214,44 +196,29 @@ proc init(com         : CommonRef,
...
-  if not com.db.getTd(blockHash, td):
-    # TODO: Is this really ok?
-    Opt.none(DifficultyInt)
-  else:
-    Opt.some(td)
-
-func needTdForHardForkDetermination(com: CommonRef): bool =
-  let t = com.forkTransitionTable.mergeForkTransitionThreshold
-  t.ttdPassed.isNone and t.number.isNone and t.ttd.isSome
+proc isBlockAfterTtd(com: CommonRef, header: Header): bool =
```
Comment (tersec) [inline nimbus/common/common.nim:212]:
```
tests/test_txpool2.nim
161:      check com.isBlockAfterTtd(blk.header)
217:      check com.isBlockAfterTtd(blk.header)
304:          check com.isBlockAfterTtd(blk.header)

nimbus/common/common.nim
383:proc isBlockAfterTtd*(com: CommonRef, header: BlockHeader): bool =
404:  if com.isBlockAfterTtd(header):
```
does this have any use?

The txpool will be able to be assumed to only ever handle PoS blocks

File: nimbus/common/hardforks.nim (hunk 2 above)
Comment (jangko) [reply nimbus/common/hardforks.nim:170]: yup, for passing test vectors

Comment (tersec) [reply nimbus/common/hardforks.nim:170]: ok

File: nimbus/evm/state.nim (hunk 1 above)
Comment (jangko) [reply nimbus/evm/state.nim:249]: yes, but this code still needed for passing test vectors.

File: nimbus/common/common.nim (hunk 3 above)
Comment (jangko) [reply nimbus/common/common.nim:212]: right, will remove it

File: nimbus/common/common.nim [hunk 4]
```diff
@@ -327,6 +327,7 @@ proc proofOfStake*(com: CommonRef, header: Header): bool =
   if com.config.mergeForkBlock.isSome:
     header.number >= com.config.mergeForkBlock.get
   else:
+   # This costly check is only executed from test suite
```
Comment (tersec) [inline nimbus/common/common.nim:330]: odd 1-space indentation?

Comment (tersec) [reply nimbus/common/common.nim:330]: the `if` branch uses 2-spaces

File: nimbus/common/common.nim (hunk 3 above)
Comment (jangko) [reply nimbus/common/common.nim:212]: https://github.com/status-im/nimbus-eth1/pull/2682/commits/5f6688a7378698abe16a496cf9b2eda78fbe0841

File: nimbus/common/common.nim (hunk 4 above)
Comment (jangko) [reply nimbus/common/common.nim]: https://github.com/status-im/nimbus-eth1/pull/2682/commits/4b3d27de2855cb1451d37b67f4e57807c513ad8b

**Pull request merged**